### PR TITLE
Enhance Confirmation Statement API to handle registered email address

### PR DESCRIPTION
**Pull request by nevans4-ch**
Various subtasks of [ECCT-714] Enhance Confirmation Statement API to handle registered email address:
* [ECCT-716] Implement a GET registered-email-address endpoint
* [ECCT-739] Update the POST and GET confirmation-statement endpoint
* [ECCT-741] Update the GET validation-status endpoint
* [ECCT-782] Convert rea call to use sdk
* [ECCT-728] Update the Swagger API for the Confirmation Statement service

[ECCT-714]: https://companieshouse.atlassian.net/browse/ECCT-714?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ECCT-716]: https://companieshouse.atlassian.net/browse/ECCT-716?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ECCT-739]: https://companieshouse.atlassian.net/browse/ECCT-739?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ECCT-741]: https://companieshouse.atlassian.net/browse/ECCT-741?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ECCT-782]: https://companieshouse.atlassian.net/browse/ECCT-782?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[ECCT-728]: https://companieshouse.atlassian.net/browse/ECCT-728?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by adeharrison-ch — CHANGES_REQUESTED** [review]

File: spec/schema.json [hunk 1]
```diff
@@ -285,6 +285,45 @@
         }
       }
     },
+    "/transactions/{transaction_id}/confirmation-statement/{confirmation_statement_id}/registered-email-address": {
```
Comment (gshivach) [inline spec/schema.json:288]: Should this path be removed? think the private get email address replaces this now

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java [hunk 2]
```diff
@@ -165,26 +170,27 @@ public ResponseEntity<Object> updateConfirmationStatement(String submissionId, C
...
             } else {
-
                 boolean isValid = isConfirmed(submissionData.getShareholdersData()) &&
                         isConfirmed(submissionData.getSicCodeData()) &&
                         isConfirmed(submissionData.getActiveOfficerDetailsData()) &&
                         isConfirmed(submissionData.getStatementOfCapitalData()) &&
                         isConfirmed(submissionData.getRegisteredOfficeAddressData()) &&
                         isConfirmed(submissionData.getPersonsSignificantControlData()) &&
                         isConfirmed(submissionData.getRegisterLocationsData()) &&
+                        isConfirmed(submissionData.getRegisteredEmailAddressData(), submissionData.getMadeUpToDate()) &&
```
Comment (gshivach) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:188]: We should ideally only check for the rea status if ecct is enabled

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java [hunk 3]
```diff
@@ -193,20 +199,35 @@ public ValidationStatusResponse isValid(String submissionId) throws SubmissionNo
...
+                    sectionData.getSectionStatus() == SectionStatus.INITIAL_FILING);
+        }
+
+        return true;
+    }
+
+    private boolean isEcctEnabled(LocalDate madeUpToDate) {
+        var ecctStartDate = LocalDate.parse(ecctStartDateStr, DATE_TIME_FORMATTER);
+
+        return isBeforeOrEqual(madeUpToDate, ecctStartDate);
```
Comment (gshivach) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:228]: madeupdate should be on or after ecctstartdate. Maybe Javadoc to clarify this logic.

File: pom.xml [hunk 4]
```diff
@@ -14,6 +14,8 @@
     <properties>
         <java.version>11</java.version>
         <spring-boot-dependencies.version>2.5.6</spring-boot-dependencies.version>
+        <api-sdk-java.version>4.3.22</api-sdk-java.version>
```
[the remark this answers is not in the document]
Comment (pparlett) [reply pom.xml]: done

File: pom.xml [hunk 5]
```diff
@@ -14,6 +14,8 @@
...
+        <private-api-sdk-java.version>2.0.286</private-api-sdk-java.version>
```
[the remark this answers is not in the document]
Comment (pparlett) [reply pom.xml]: done

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java (hunk 2 above)
Comment (gshivach) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:208]: agree

File: spec/schema.json (hunk 1 above)
Comment (tcbaileych) [reply spec/schema.json]: done

File: spec/schema.json [hunk 6]
```diff
@@ -589,6 +667,17 @@
             }
           }
         },
+        "registered_email_address_data": {
```
[the remark this answers is not in the document]
Comment (tcbaileych) [reply spec/schema.json]: fixed

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java (hunk 3 above)
Comment (gshivach) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:246]: Still looks wrong

Comment (pparlett) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:246]: I agree it is wrong, apologies for my confusion! I need to look at the unit test for this logic...

Comment (pparlett) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:246]: It looks a bit strange but it is actually correct, the parameter order of isBeforeOrEqual may be confusing.  I have added a Javadoc for that method in a new commit to clarify. (I have also checked how the original author of that method was using it, see call on line 215:

```
isBeforeOrEqual(localDateNow.get(), submissionData.getMadeUpToDate());
```

**Comment by tcbaileych** [comment]
Merge conflict between ecct-165 version `.getShareholdersData()` and main version `.getShareholderData`:
Decided to keep `.getShareholdersData()` as the method is used throughout the repository and has unit test cases included.

**Review by adeharrison-ch — APPROVED** [review]
LGTM

**Pull request merged**